File: src/main/java/com/aquasertao/api/modules/operational/services/ReportService.java
package com.aquasertao.api.modules.operational.services;

import com.aquasertao.api.modules.billing.repositories.SaaSPlanRepository;
import com.aquasertao.api.modules.billing.repositories.SubscriptionRepository;
import com.aquasertao.api.modules.operational.dtos.ReportSummaryDTO;
import com.aquasertao.api.modules.operational.models.BiometricsRecord;
import com.aquasertao.api.modules.operational.models.MortalityRecord;
import com.aquasertao.api.modules.operational.models.Tank;
import com.aquasertao.api.modules.operational.repositories.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ReportService {

    private final TankRepository tankRepository;
    private final BiometricsRecordRepository biometricsRecordRepository;
    private final MortalityRecordRepository mortalityRecordRepository;
    private final FeedingRecordRepository feedingRecordRepository;
    private final HarvestRepository harvestRepository;
    private final SubscriptionRepository subscriptionRepository;
    private final SaaSPlanRepository saasPlanRepository;

    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ofPattern("dd/MM/yyyy");

    public ReportSummaryDTO generateReport(UUID farmId) {

        List<Tank> tanks = tankRepository.findByFarmId(farmId);

        // ── 1. Growth history (all biometrics across all tanks) ──────────────
        List<ReportSummaryDTO.GrowthPoint> growthPoints = tanks.stream()
                .flatMap(tank -> {
                    List<BiometricsRecord> records =
                            biometricsRecordRepository.findByTankIdAndFarmIdOrderByRecordDateDesc(tank.getId(), farmId);
                    return records.stream().map(r -> ReportSummaryDTO.GrowthPoint.builder()
                            .date(r.getRecordDate().format(DATE_FMT))
                            .avgWeightG(r.getWeightG() != null ? r.getWeightG().doubleValue() : 0.0)
                            .tankName(tank.getName())
                            .build());
                })
                .sorted(Comparator.comparing(ReportSummaryDTO.GrowthPoint::getDate))
                .collect(Collectors.toList());

        // ── 2. Mortality ─────────────────────────────────────────────────────
        List<MortalityRecord> allMortality = tanks.stream()
                .flatMap(t -> mortalityRecordRepository
                        .findByTankIdAndFarmIdOrderByRecordDateDesc(t.getId(), farmId).stream())
                .collect(Collectors.toList());

        int totalMortality = allMortality.stream()
                .mapToInt(m -> m.getQuantity() != null ? m.getQuantity() : 0)
                .sum();

        long totalStocked = tanks.stream()
                .mapToLong(t -> t.getInitialStockingQty() != null ? t.getInitialStockingQty() : 0)
                .sum();

        double mortalityRate = totalStocked > 0
                ? Math.round((double) totalMortality / totalStocked * 10000.0) / 100.0
                : 0.0;

        List<ReportSummaryDTO.MortalityPoint> mortalityPoints = allMortality.stream()
                .map(m -> ReportSummaryDTO.MortalityPoint.builder()
                        .date(m.getRecordDate().format(DateTimeFormatter.ofPattern("dd/MM/yyyy")))
                        .count(m.getQuantity())
                        .cause(m.getCause())
                        .build())
                .sorted(Comparator.comparing(ReportSummaryDTO.MortalityPoint::getDate))
                .collect(Collectors.toList());

        // ── 3. Feeding / Ração ────────────────────────────────────────────────
        // Group feeding records by date (last 30 days from all records)
        var allFeedings = feedingRecordRepository.findAll().stream()
                .filter(f -> f.getFarmId().equals(farmId))
                .collect(Collectors.toList());

        BigDecimal totalFeedKg = allFeedings.stream()
                .map(f -> f.getQuantity() != null ? f.getQuantity() : BigDecimal.ZERO)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal totalFeedCost = allFeedings.stream()
                .map(f -> {
                    BigDecimal qty = f.getQuantity() != null ? f.getQuantity() : BigDecimal.ZERO;
                    BigDecimal cost = f.getUnitCost() != null ? f.getUnitCost() : BigDecimal.ZERO;
                    return qty.multiply(cost);
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Group by date string
        Map<String, BigDecimal[]> feedByDate = new LinkedHashMap<>();
        allFeedings.forEach(f -> {
            String day = f.getFeedingTime().format(DATE_FMT);
            feedByDate.computeIfAbsent(day, k -> new BigDecimal[]{BigDecimal.ZERO, BigDecimal.ZERO});
            BigDecimal qty = f.getQuantity() != null ? f.getQuantity() : BigDecimal.ZERO;
            BigDecimal cost = f.getUnitCost() != null ? f.getUnitCost() : BigDecimal.ZERO;
            feedByDate.get(day)[0] = feedByDate.get(day)[0].add(qty);
            feedByDate.get(day)[1] = feedByDate.get(day)[1].add(qty.multiply(cost));
        });

        List<ReportSummaryDTO.FeedingPoint> feedingPoints = feedByDate.entrySet().stream()
                .map(e -> ReportSummaryDTO.FeedingPoint.builder()
                        .date(e.getKey())
                        .quantityKg(e.getValue()[0])
                        .cost(e.getValue()[1])
                        .build())
                .collect(Collectors.toList());

        // ── 4. Harvest Forecasts ─────────────────────────────────────────────
        List<ReportSummaryDTO.HarvestForecast> forecasts = tanks.stream()
                .filter(t -> t.getNextHarvestDate() != null)
                .map(t -> {
                    int fishCount = t.getInitialStockingQty() != null ? t.getInitialStockingQty() : 0;
                    int mortality = t.getMortalityCount() != null ? t.getMortalityCount() : 0;
                    int alive = Math.max(0, fishCount - mortality);
                    double avgWeightKg = t.getAverageWeightG() != null ? t.getAverageWeightG() / 1000.0 : 0.0;
                    return ReportSummaryDTO.HarvestForecast.builder()
                            .tankName(t.getName())
                            .expectedDate(t.getNextHarvestDate().format(DATE_FMT))
                            .estimatedFishCount(alive)
                            .estimatedWeightKg(Math.round(alive * avgWeightKg * 100.0) / 100.0)
                            .build();
                })
                .collect(Collectors.toList());

        return ReportSummaryDTO.builder()
                .growthHistory(growthPoints)
                .totalMortality(totalMortality)
                .mortalityRate(mortalityRate)
                .mortalityHistory(mortalityPoints)
                .totalFeedKg(totalFeedKg)
                .totalFeedCost(totalFeedCost)
                .feedingHistory(feedingPoints)
                .harvestForecasts(forecasts)
                .build();
    }

    // ── Plan guard: returns max tanks allowed for this farm ──────────────────
    public int getMaxTanksAllowed(UUID farmId) {
        return subscriptionRepository.findByFarmIdAndStatus(farmId, "ACTIVE")
                .map(sub -> saasPlanRepository.findById(sub.getPlanId())
                        .map(plan -> plan.getMaxTanks())
                        .orElse(1))
                .orElse(1); // default: FREE plan = 1 tank
    }
}

File: frontend_flutter/lib/features/reports/data/report_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

// ─────────────────────────────────────────────────────────────────────────────
// Data models
// ─────────────────────────────────────────────────────────────────────────────

class GrowthPoint {
  final String date;
  final double avgWeightG;
  final String tankName;
  GrowthPoint({required this.date, required this.avgWeightG, required this.tankName});
  factory GrowthPoint.fromJson(Map<String, dynamic> j) => GrowthPoint(
        date: j['date'] ?? '',
        avgWeightG: (j['avgWeightG'] ?? 0).toDouble(),
        tankName: j['tankName'] ?? '',
      );
}

class MortalityPoint {
  final String date;
  final int count;
  final String? cause;
  MortalityPoint({required this.date, required this.count, this.cause});
  factory MortalityPoint.fromJson(Map<String, dynamic> j) => MortalityPoint(
        date: j['date'] ?? '',
        count: j['count'] ?? 0,
        cause: j['cause'],
      );
}

class FeedingPoint {
  final String date;
  final double quantityKg;
  final double cost;
  FeedingPoint({required this.date, required this.quantityKg, required this.cost});
  factory FeedingPoint.fromJson(Map<String, dynamic> j) => FeedingPoint(
        date: j['date'] ?? '',
        quantityKg: (j['quantityKg'] ?? 0).toDouble(),
        cost: (j['cost'] ?? 0).toDouble(),
      );
}

class HarvestForecast {
  final String tankName;
  final String expectedDate;
  final int estimatedFishCount;
  final double estimatedWeightKg;
  HarvestForecast({
    required this.tankName,
    required this.expectedDate,
    required this.estimatedFishCount,
    required this.estimatedWeightKg,
  });
  factory HarvestForecast.fromJson(Map<String, dynamic> j) => HarvestForecast(
        tankName: j['tankName'] ?? '',
        expectedDate: j['expectedDate'] ?? '',
        estimatedFishCount: j['estimatedFishCount'] ?? 0,
        estimatedWeightKg: (j['estimatedWeightKg'] ?? 0).toDouble(),
      );
}

class ReportSummary {
  final List<GrowthPoint> growthHistory;
  final int totalMortality;
  final double mortalityRate;
  final List<MortalityPoint> mortalityHistory;
  final double totalFeedKg;
  final double totalFeedCost;
  final List<FeedingPoint> feedingHistory;
  final List<HarvestForecast> harvestForecasts;

  ReportSummary({
    required this.growthHistory,
    required this.totalMortality,
    required this.mortalityRate,
    required this.mortalityHistory,
    required this.totalFeedKg,
    required this.totalFeedCost,
    required this.feedingHistory,
    required this.harvestForecasts,
  });

  factory ReportSummary.fromJson(Map<String, dynamic> j) => ReportSummary(
        growthHistory: (j['growthHistory'] as List? ?? []).map((e) => GrowthPoint.fromJson(e)).toList(),
        totalMortality: j['totalMortality'] ?? 0,
        mortalityRate: (j['mortalityRate'] ?? 0).toDouble(),
        mortalityHistory: (j['mortalityHistory'] as List? ?? []).map((e) => MortalityPoint.fromJson(e)).toList(),
        totalFeedKg: (j['totalFeedKg'] ?? 0).toDouble(),
        totalFeedCost: (j['totalFeedCost'] ?? 0).toDouble(),
        feedingHistory: (j['feedingHistory'] as List? ?? []).map((e) => FeedingPoint.fromJson(e)).toList(),
        harvestForecasts: (j['harvestForecasts'] as List? ?? []).map((e) => HarvestForecast.fromJson(e)).toList(),
      );
}

// ─────────────────────────────────────────────────────────────────────────────
// Service
// ─────────────────────────────────────────────────────────────────────────────

class ReportService {
  final Dio _dio;

  ReportService({required Dio dio}) : _dio = dio;

  Future<ReportSummary> fetchSummary(String farmId) async {
    final response = await _dio.get(
      '/api/reports/summary',
      queryParameters: {'farmId': farmId},
    );
    return ReportSummary.fromJson(response.data as Map<String, dynamic>);
  }

  Future<int> fetchPlanLimit(String farmId) async {
    final response = await _dio.get(
      '/api/reports/plan-limit',
      queryParameters: {'farmId': farmId},
    );
    return (response.data['maxTanks'] ?? 1) as int;
  }
}
